### PR TITLE
Revamped shuttle creation through the power of science!

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Improved the rapid shuttle designator to allow for weirdly shaped, and non-pressurised shuttles.

There are now new actions:
 - Designate Area (Designates the area you are in, same as before)
 - Designate Turf (Designates the turf you are above)
 - Clear Area (Resets the marked area)
 - Clear Turf (Removes the turf you are above)
 - Select Docking AIrlock (Finishes the shuttle by marking the airlock you are above as the docking port.)

The drone can only move up to 12 units away from the source, if designating larger shuttles / longer shuttles you can designate one side, move over and then designate the next part.

This makes shuttle creation a lot easier and I'll get less ahelps that need me to fix things.

image

image

image

image

The creator eye cannot see mobs, so you cannot use it as a cheap wallhack to see where people are perfoming nefarious acts.

## Why It's Good For The Game

Makes a dodgey shitcodey tool, slightly less dodgey (but not necessarilly less shitcodey)
Moves different elements of the tool into different files, heavilly improving readability.

## Changelog
:cl:
refactor: Refactored shuttle designation
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Bug: leaving camera doesn't remove the image

**Comment by PowerfulBacon** [comment]
Done 
image

Ready for review

File: code/modules/shuttle/shuttle_creation/shuttle_creator.dm [hunk 1]
```diff
@@ -24,19 +27,43 @@ GLOBAL_LIST_EMPTY(custom_shuttle_machines)		//Machines that require updating (He
...
+/obj/item/shuttle_creator/Initialize()
+	. = ..()
+	internal_shuttle_creator = new()
+	internal_shuttle_creator.owner_rsd = src
+	overlay_holder = new()
+
+/obj/item/shuttle_creator/Destroy()
+	. = ..()
+	if(internal_shuttle_creator)
+		qdel(internal_shuttle_creator)
```
Comment (ike709) [inline code/modules/shuttle/shuttle_creation/shuttle_creator.dm]:
```suggestion
		QDEL_NULL(internal_shuttle_creator)
```

File: code/modules/shuttle/shuttle_creation/shuttle_creator.dm [hunk 2]
```diff
@@ -24,19 +27,43 @@ GLOBAL_LIST_EMPTY(custom_shuttle_machines)		//Machines that require updating (He
...
+	if(overlay_holder)
+		qdel(overlay_holder)
```
Comment (ike709) [inline code/modules/shuttle/shuttle_creation/shuttle_creator.dm]:
```suggestion
		QDEL_NULL(overlay_holder)
```

File: code/modules/shuttle/shuttle_creation/shuttle_creator.dm [hunk 3]
```diff
@@ -24,19 +27,43 @@ GLOBAL_LIST_EMPTY(custom_shuttle_machines)		//Machines that require updating (He
...
+		QDEL_NULL(internal_shuttle_creator)
+	if(overlay_holder)
+		QDEL_NULL(overlay_holder)
+
 /obj/item/shuttle_creator/attack_self(mob/user)
 	..()
 	if(linkedShuttleId)
 		return
-	if(GLOB.custom_shuttle_count > CUSTOM_SHUTTLE_LIMIT)
+	if(GLOB.custom_shuttle_count > CUSTOM_SHUTTLE_LIMIT && override_max_shuttles)
```
Comment (ike709) [inline code/modules/shuttle/shuttle_creation/shuttle_creator.dm]:
```suggestion
	if(GLOB.custom_shuttle_count > CUSTOM_SHUTTLE_LIMIT && !override_max_shuttles)
```

Unless I'm misreading?

Comment (PowerfulBacon) [reply code/modules/shuttle/shuttle_creation/shuttle_creator.dm]: No, you are correct

File: code/modules/shuttle/shuttle_creation/shuttle_creator.dm [hunk 4]
```diff
@@ -24,19 +27,43 @@ GLOBAL_LIST_EMPTY(custom_shuttle_machines)		//Machines that require updating (He
...
+/obj/item/shuttle_creator/Initialize()
+	. = ..()
+	internal_shuttle_creator = new()
+	internal_shuttle_creator.owner_rsd = src
+	overlay_holder = new()
+
+/obj/item/shuttle_creator/Destroy()
+	. = ..()
+	if(internal_shuttle_creator)
+		QDEL_NULL(internal_shuttle_creator)
```
Comment (ike709) [inline code/modules/shuttle/shuttle_creation/shuttle_creator.dm]:
```suggestion
	if(internal_shuttle_creator)
		internal_shuttle_creator.owner_rsd = null
		QDEL_NULL(internal_shuttle_creator)
```

**Comment by MCterra10** [comment]
I assume you can't designate areas that are part of the station, correct?

**Comment by PowerfulBacon** [comment]
> I assume you can't designate areas that are part of the station, correct?

Of course, for both area and turf designation the same system is still used.

**Review by ike709 — APPROVED** [review]

File: code/modules/shuttle/shuttle_creation/shuttle_creator.dm [hunk 5]
```diff
@@ -268,33 +292,12 @@ GLOBAL_LIST_EMPTY(custom_shuttle_machines)		//Machines that require updating (He
...
-	to_chat(user, "<span class='notice'>Your current area was logged into the [src], select an airlock to act as the docking point.</span>")
-	return TRUE
-
+//Checks an area to ensure that the turfs provided are valid to be made into a shuttle
 /obj/item/shuttle_creator/proc/check_area(list/turfs)
 	if(!turfs)
 		to_chat(usr, "<span class='warning'>Shuttles must be created in an airtight space, ensure that the shuttle is airtight, including corners.</span>")
 		return FALSE
-	if(turfs.len > SHUTTLE_CREATOR_MAX_SIZE)
+	if(turfs.len + loggedTurfs.len > SHUTTLE_CREATOR_MAX_SIZE)
```
Comment (PowerfulBacon) [inline code/modules/shuttle/shuttle_creation/shuttle_creator.dm:300]: this breaks it completely by doubling the area if in final check

**Pull request merged**